terraform/dns/main.tf

```diff
@@ -0,0 +1,229 @@
+resource "aws_ses_receipt_rule" "main" {
+  depends_on    = [aws_s3_bucket_policy.emails]
+  name          = "main"
+  rule_set_name = aws_ses_receipt_rule_set.main.rule_set_name
+  enabled       = true
+  s3_action {
+    bucket_name = "${var.domain}-emails"
+    // If you have more than one action, you can decide the order they're applied in by changing this attribute.
+    position = 1
+  }
+}
+
+// Now that we have set up our rule set with all the rules we need, make the rule set active.
+resource "aws_ses_active_receipt_rule_set" "main" {
+  rule_set_name = aws_ses_receipt_rule_set.main.rule_set_name
+}
+
+// The last thing to do is to set up the MX record for our domain.
+// You can see that we've set up an entire mail inbox for our domain, including verification, completely within the same flow in the same Terraform script.
+resource "dnsimple_zone_record" "mx_record" {
+  zone_name = var.domain
+  name      = ""
+  type      = "MX"
+  ttl       = 600
+  priority  = 10
+  value     = "inbound-smtp.us-east-1.amazonaws.com"
+}
+
+// Now that we have some data (emails stored in our S3 bucket), we'll create an API to expose this information for our app (which we'll set up later). This is a common pattern for many projects and part of most infrastructure. We'll use EC2 instances to host our API server, and see how we can quickly set up the DNS plumbing to get our service up and running rapidly, right here in Terraform.
+
+// First, we need to create a role that will allow our instances to read the emails in our bucket.
+resource "aws_iam_role" "api" {
+  name = "api"
+
+  inline_policy {
+    name = "main"
+    policy = jsonencode(templatefile("iam/api-permissions.json", {
+      domain = var.domain,
+    }))
+  }
+
+  // Allow our EC2 instances (which we'll create later) to assume this role.
+  assume_role_policy = jsonencode(file("iam/api-role-assume.json"))
+}
+
+// We'll create an instance profile so that instances can assume the role we've just created. Learn more about them at https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use_switch-role-ec2_instance-profiles.html.
+resource "aws_iam_instance_profile" "api" {
+  name = "api"
+  role = aws_iam_role.api.name
+}
+
+// Now, we can create the instances and deploy our API service. We'll use a simple mechanism to deploy and run the service, which is to simply copy the Python code into our cloud-init script and run it immediately and indefinitely when our instance starts up. You can learn more about cloud-init at https://cloud-init.io/.
+// As our API server listens on port 80 (plaintext HTTP), ensure this server can be reached publicly over port 80, so adjust any VPC, subnet, and NSG attributes as necessary. WARNING: Because this is unencrypted, for the purposes of this demo, don't send any private or sensitive information via email to this domain!
+// Also, if you'd like to set up SSH access, make sure to provide `key_name`.
+resource "aws_instance" "api-virginia" {
```

Comment:
Missing an SG that allows HTTP traffic. Same for the SYD instance. You mention this as a comment, but it might be missed by the user and also requires them to come out of the demo and wrangle with the AWS console.